chainer/exporters/caffe.py

```diff
@@ -0,0 +1,428 @@
+import collections
+
+import heapq
+import numpy
+import os
+import six
+
+import chainer
+from chainer import function
+from chainer import function_node
+from chainer.links.caffe.protobuf3 import caffe_pb2 as caffe_pb
+from chainer import variable
+
+
+_function_types = (function.Function, function_node.FunctionNode)
+
+
+def _add_blob(layer, shape, data):
+    # The following part is ridiculously slow!!
+    # TODO(who): Replace with C++ extension call
+    blob = layer.blobs.add()
+    blob.shape.dim[:] = shape
+    blob.data[:] = data.flatten()
+
+
+def _dump_graph(outputs):
+    fan_out = collections.defaultdict(int)
+    cand_funcs = []
+
+    def add_cand_to_check(cands):
+        for cand in cands:
+            x = cand.creator
+            if x is None:
+                continue
+            if x not in fan_out:
+                # `len(fan_out)` is in order to avoid comparing `x`
+                heapq.heappush(cand_funcs, (-x.rank, len(fan_out), x))
+            fan_out[x] += 1
+
+    add_cand_to_check(outputs)
+    while cand_funcs:
+        _, _, func = heapq.heappop(cand_funcs)
+        assert isinstance(func, _function_types)
+        add_cand_to_check(func.inputs)
+
+    ret = []
+    cand_funcs = []
+    seen_set = set()
+
+    def add_cand(cands):
+        cands = [cand.creator for cand in cands if cand.creator is not None]
+        for x in cands:
+            if x in seen_set:
+                continue
+            order = 1
+            if fan_out[x] == 1 and len(cands) == 1:
+                order = -len(seen_set)
+            # Negate since heapq is min-heap
+            # `len(seen_set)` is in order to avoid comparing `x`
+            heapq.heappush(cand_funcs, (order, -x.rank, -len(seen_set), x))
+            seen_set.add(x)
+
+    add_cand(outputs)
+    while cand_funcs:
+        _, _, _, func = heapq.heappop(cand_funcs)
+        ret.append(func)
+        add_cand(func.inputs)
+
+    return ret[::-1]
+
+
+class _RetrieveAsCaffeModel(object):
+
+    debug = False
+
+    def __init__(self, prototxt, caffemodel=None):
+        self.caffemodel = caffemodel
+        self.prototxt = prototxt
+        # key:string, val:dict(key: func, val: index)
+        self.naming_map = collections.defaultdict(dict)
+
+    def _get_layer_name(self, layer):
+        """Generate layer name like "Convolution2DFunction-10-2".
+
+        The first number means rank of the layer (depth from the top),
+        and the second number is for preventing duplication
+        (different layer objects can have same rank)
+
+        :param layer: Function object
+        :return: string unique to functions
```

Comment:
The docstring format is different from others (this docstring is not exposed to users, though).